options/rtdl/generic/main.cpp

```diff
@@ -385,16 +402,22 @@ void *__dlapi_open(const char *file, int local) {
 }
 
 extern "C" [[ gnu::visibility("default") ]]
-void *__dlapi_resolve(void *handle, const char *string) {
+void *__dlapi_resolve(void *handle, const char *string, void *returnAddress) {
 	if (logDlCalls)
 		mlibc::infoLogger() << "rtdl: __dlapi_resolve(" << handle << ", " << string << ")" << frg::endlog;
 
-	__ensure(handle != RTLD_NEXT);
-
 	frg::optional<ObjectSymbol> target;
 
 	if (handle == RTLD_DEFAULT) {
 		target = Scope::resolveWholeScope(globalScope.get(), string, 0);
+	} else if (handle == RTLD_NEXT) {
+		SharedObject *origin = initialRepository->findCaller(returnAddress);
+		if (!origin) {
+			mlibc::panicLogger() << "rtdl: unable to determine calling object of dlsym "
+				<< "(ra = " << returnAddress << ")" << frg::endlog;
+		}
+
+		target = Scope::resolveNext(globalScope.get(), string, origin);
```

Comment:
The local scope is not handled here (yet)?

Reply:
Yeah, I guess we just use the object's scope when that is implemented